### DISCUSSION TITLE
Custom Damage Events

### DISCUSSION THREAD

**Discussion by James103** [gnembon/fabric-carpet#649] [Ideas]
_Moved from #251 in order to stimulate feedback on the topic_

Even as of Carpet mod 1.4.20 for Minecraft 1.16.4, we still don't have a way to deal an arbitrary amount of any damage type to any entity, on behalf of any entity. I suggest that there be a function to do exactly this:

## `modify(e, 'damage', amount, source, source_entity)`
Deals `amount` damage of type `source` to the target entity on behalf of `source_entity` and calls the appropriate `__on_damaged` (or `__on_player_takes_damage`) event if available. All vanilla damage types are supported (with corresponding effects and death messages), as well as custom damage types. `source_entity` is optional and defaults to `null` (to denote no source entity).
If the damage is nullified because the entity has already taken damage, the event is cancelled and no callbacks are made.

## Example use cases:
You take constant fire damage while in the Nether:
`modify(player, 'damage', 1, 'fire', null)` or `modify(player, 'damage', 1, 'fire')`

Kill yourself and destroy all your equipped armor:
`modify(player, 'damage', 9999, 'mob', player)`

Instantly die once your hunger hits 0:
`modify(player, 'damage', player~'health', 'starve', null)` or `modify(player, 'damage', player~'health', 'starve')`

Reflect all damage you take back onto the attacker...:
```
__on_player_takes_damage(player, amount, source, source_entity) -> (
	modify(source_entity, 'damage', amount, source, player);
);
```

...and vice versa:
```
__on_player_deals_damage(player, amount, entity) -> (
	modify(player, 'damage', amount, 'thorns', entity);
);
```

Villagers attack you whenever you trade with them:
```
__on_player_trades(player, entity, buy_left, buy_right, sell) -> (
	modify(player, 'damage', 1, 'mob', entity);
);
```

A custom damage type example based on the above example
```
__on_player_trades(player, entity, buy_left, buy_right, sell) -> (
	modify(player, 'damage', 1, 'thorns_trade', entity);
);
```
> Player was killed trying to trade with Farmer

As you can see, by having custom damage events, there are a lot more possibilities for challenge apps and datapacks that can damage the player and/or entities in a whole manner of different ways. Any thoughts on this?

**Comment by replaceitem** [comment]
As per [scarpet#179](https://github.com/gnembon/scarpet/issues/179), a knockback direction could be added in the form of a vector.
A potential issue with dealing damage inside the event could be, if a player hits another player, it would recursively deal damage back and forth, which when the players are invincible (e.g. resistance effect given by commands) would freeze the server.

**Reply by James103** [reply]
A workaround may be to have a scoreboard flag or tag that is set when the event is called and is reset every tick. In other words,
```
__on_player_deals_damage(player, amount, entity) -> (
	if(scoreboard('P_DamageTaken', player) >= amount, return());
	scoreboard('P_DamageTaken', player, amount);
	modify(player, 'damage', amount, 'thorns', entity);
);

__on_tick() -> (
	for(entity_list('*'), scoreboard('P_DamageTaken', _, 0));
);

**Reply by replaceitem** [reply]
Yeah, something like that should be done, but im wondering if some recursion protection should be done in carpet. Is there any other event in scarpet that can be triggered with a function leading to a similar case?

**Reply by zane49er2** [reply]
I think it would be reasonable to check for damage_time by default and cancel the effect if it is used too rapidly unless you set damage time and schedule the damage to the next tick

There are already ways you can use scarlet to overload the server, with particles or math, so I don't see a need for more protection than that

**Reply by zane49er2** [reply]
I would also like to note that this function should ideally turn the mob red, cause screen shake for the player, or play the appropriate sounds

**Reply by replaceitem** [reply]
Yeah overloading the server can of course be done, but in that case it would be recursion, which freezes it, and makes commands not even be processed, even console input wont work to stop the script, you would habe to kill the server. And im not sure if something like that should be able to do that... Datapack funktions use the command chain length gamerule as recursion protection for example, but not sure if scarpet should have something like that too...
And with the red overlay, shaking screen and sounds, thats all handled by the client, it would be harder not to include that

**Reply by replaceitem** [reply]
client* its handled by the client, i typed without thinking

**Comment by mathgeniuszach** [comment]
Definitely necessary for making custom hitscan weapons. I'm actually surprised scarpet doesn't have this yet.

**Comment by zane49er2** [comment]
When revisiting this recently I had a second suggestion for an argument to add (I was the one to suggest knockback vect):
With block-breaking, you can pass in a tool. This could also be helpful for the damage function, to wear out swords, or take looting into consideration.

**Reply by Ghoulboy78** [reply]
Well that could just be done by looking at the `source_entity` and modifying its tools etc.
